lib/command.coffee

```diff
@@ -70,12 +70,25 @@ module.exports = class Command
 					return callback?(error) if error?
 
 					try
-						@action(params, parsedOptions, callback)
+						actionPromise = @action(params, parsedOptions, callback)
 					catch error
 						return callback?(error)
 
-					# Means the user is not declaring the callback
-					return callback?() if @action.length < 3
+					if callback?
+						if actionPromise?.then?
+							# Asynchronous action function
+							actionPromise.then(
+								# the value of a fulfilled promise is intentionally ignored because users are expected
+								# to call the callback themselves, and also for backwards compatibility.
+								(_value) => callback() if @action.length < 3  # the user hasn't declared the callback
+
+								# call the callback if the promise is rejected (e.g. if an async action function throws
+								# an error), whether or not the user declared a callback. This is consistent with the
+								# existing implementation for synchronous action functions that throw an error.
+								(_error) -> callback _error)
```

Comment:
@zvin, thanks for the review. I've made a small change on this line after your approval. Previously it was:
 `(_error) -> callback _error if _error)`
I have deleted `if _error` because it was redundant. The tests pass all the same.

> _Approved as the tests work but this execute method mixes promises, callbacks and optional parameters and is highly unreadable._

Yeah, it's a bit messy, right? There's the implementation of the `execute` method, but actually I think there is also a design issue: Capitano allows users to write an `action` method that takes a callback and returns a promise -- this is implicitly endorsed by Capitano's documentation quoted in issue #67. For example, `balena-cli` prefixes the `action` method with the `async` keyword (at least for the `push` command), which means it returns a promise and this led to the issue of a thrown error being uncaught. But I don't know we can do much about it without breaking backwards compatibility. If we were to break backwards compatibility (I don't think we should!), then we could do something like making callbacks and promises mutually exclusive. If the action method returned a promise, it should not be allowed to declare a callback and should use the promise instead of the callback to report on success or failure. (CC: @jviotti)

Reply:
`(_error) -> callback _error` is equivalent to `callback`

Reply:
@zvin, you know what, you are right! :-) Neat. I have updated the code, and also deleted the `return` keyword mentioned in your other comment.